docs/basics/upgradeability.md

```diff
@@ -122,6 +131,172 @@ Using this pattern, you can introduce other message to your proxy contract.
 Any messages that are not matched in the proxy contract 
 will be forwarded to the specified contract address.
 
+## Delegating execution to foreign Contract Code with `delegate_call`
+
+Similar to proxy-forwarding we can delegate execution to another code hash uploaded on-chain.
+
+### Properties 
+
+- Delegates any call that does not match a selector of itself to another contract.
+- Code is required to uploaded on-chain, but is not required to be instantiated.
+- State is stored in the storage of the original contract which submits the call.
+- Storage layout must be identical between both contract codes.
+
+```
+User ---- tx ---> Contract ----------> Code_v0
+                     ^                     |
+                     |                     |
+                     ⌊_____________________⌋
+                  Storage update is propagated
+```
+
+### Example
+
+Suppose we have the defined of the caller contract as following:
+
+```rust
+#[ink(storage)]
+pub struct Delegator {
+    addresses: Mapping<AccountId, i32, ManualKey<0x23>>,
+    counter: i32,
+}
+```
+
+Then let's define two messages that separately calls to update `addresses` and `counter` separately:
+
+```rust
+/// Increment the current value using delegate call.
+#[ink(message)]
+pub fn inc_delegate(&mut self, hash: Hash) {
+    let selector = ink::selector_bytes!("inc");
+    let _ = build_call::<DefaultEnvironment>()
+        .delegate(hash)
+        // We specify `set_tail_call(true)` to use the delegatee last memory frame
+        // as the end of the execution cycle.
+        // So any mutations to `Packed` types, made by delegatee,
+        // will be flushed to storage.
+        //
+        // If we don't specify this flag.
+        // The storage state before the delegate call will be flushed to storage instead.
+        // See https://substrate.stackexchange.com/questions/3336/i-found-set-allow-reentry-may-have-some-problems/3352#3352
+        .call_flags(CallFlags::default().set_tail_call(true))
```

Comment:
This is very unintuitive, we need a better way. 

It believe it might work by setting the receiver to `&self` instead of `&mut self`.

Reply:
Please elaborate. AFAIK, this has nothing to do with the received. It's about the execution stack. If tail is not set, any changes done in the delegator will be popped and discarded.

Reply:
The [docs for `set_tail_call`](https://github.com/paritytech/ink/blob/27407b5f59f24233cddc3cea59cc2213ba2c6674/crates/env/src/backend.rs#L107) say:
```
/// This makes the current contract completely transparent to its caller by replacing
/// this contracts potential output with the callee ones. Any code after the contract
/// calls has been invoked can be safely considered unreachable.
```
So, if it is *not* set then execution resumes in the calling contract after the delegate call.

And if you use `&mut self` as the receiver then the current state of the calling contract at the end of the execution [will be persisted](https://github.com/paritytech/ink/blob/5fb5477c2021a478d9eddbcb624c00e4b35cdea8/crates/ink/codegen/src/generator/dispatch.rs#L837).

Therefore, by changing to `&self` for the receiver it does not persist, therefore fixing the specific issue of "the storage state before the delegate call will be flushed to storage instead." which is mentioned in the comment.

I have tested this using the [delegator example](https://github.com/paritytech/ink/blob/cd0670a8ee2ecfa059735303b0f9b7539780969d/integration-tests/upgradeable-contracts/delegator/lib.rs#L59)